Ingress/controllers/nginx-third-party/controller.go

```diff
@@ -0,0 +1,316 @@
+		lbInfo: lbInfo,
+	}
+	lbc.ingQueue = NewTaskQueue(lbc.syncIngress)
+	lbc.configQueue = NewTaskQueue(lbc.syncConfig)
+
+	lbc.ngx = nginx.NewManager(kubeClient, defaultSvc, customErrorSvc, namespace, lbc.recorder)
+
+	// Ingress watch handlers
+	pathHandlers := framework.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			addIng := obj.(*extensions.Ingress)
+			lbc.recorder.Eventf(addIng, "ADD", addIng.Name)
+			lbc.ingQueue.enqueue(obj)
+		},
+		DeleteFunc: lbc.ingQueue.enqueue,
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				glog.V(2).Infof("Ingress %v changed, syncing", cur.(*extensions.Ingress).Name)
+			}
+			lbc.ingQueue.enqueue(cur)
+		},
+	}
+	lbc.ingLister.Store, lbc.ingController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc:  ingressListFunc(lbc.client, namespace),
+			WatchFunc: ingressWatchFunc(lbc.client, namespace),
+		},
+		&extensions.Ingress{}, resyncPeriod, pathHandlers)
+
+	// Config watch handlers
+	configHandlers := framework.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				glog.V(2).Infof("nginx rc changed, syncing")
+			}
+			lbc.configQueue.enqueue(cur)
+		},
+	}
+
+	lbc.configLister.Store, lbc.configController = framework.NewInformer(
+		&cache.ListWatch{
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return kubeClient.
+					Get().
+					Prefix("watch").
+					Namespace(lbInfo.RCNamespace).
+					Resource("resource").
+					Name(lbInfo.RCName).
+					Watch()
+			},
+		},
+		&api.ReplicationController{}, resyncPeriod, configHandlers)
+
+	return &lbc, nil
+}
+
+func ingressListFunc(c *client.Client, ns string) func() (runtime.Object, error) {
+	return func() (runtime.Object, error) {
+		return c.Extensions().Ingress(ns).List(labels.Everything(), fields.Everything())
+	}
+}
+
+func ingressWatchFunc(c *client.Client, ns string) func(options api.ListOptions) (watch.Interface, error) {
+	return func(options api.ListOptions) (watch.Interface, error) {
+		return c.Extensions().Ingress(ns).Watch(
+			labels.Everything(), fields.Everything(), options)
+	}
+}
+
+// syncIngress manages Ingress create/updates/deletes.
+func (lbc *loadBalancerController) syncIngress(key string) {
+	glog.V(2).Infof("Syncing Ingress %v", key)
+
+	obj, ingExists, err := lbc.ingLister.Store.GetByKey(key)
+	if err != nil {
+		lbc.ingQueue.requeue(key, err)
+		return
+	}
+
+	if !ingExists {
+		return
+	}
+
+	// No need to reload nginx
+	// Instead we invoke a POST /update-ingress to update the Ingress rules to use
+	ingList := lbc.ingLister.Store.List()
+	lbc.ngx.SyncIngress(ingList)
+
+	ing := *obj.(*extensions.Ingress)
+	if lbc.updateIngressStatus(ing); err != nil {
+		lbc.recorder.Eventf(&ing, "Status", "%v", err)
+		lbc.ingQueue.requeue(key, err)
+	}
+	return
+}
+
+// syncIngress manages Ingress create/updates/deletes.
+func (lbc *loadBalancerController) syncConfig(key string) {
```

Comment:
a controller loop should never drop updates. Say svc 1 gets a new endpoint, so you get a watch notification, and then svc1 never changes. Now assume you return from an error (like ReadConfig) without requeuing yourself. That new endpoint will get dropped forever. Suggest requeuing and logging all errors. Only time to not requeue would be something like a delete failing with 404, or create failing with resource exists.